src/search/FindInFiles.js

```diff
@@ -830,6 +856,56 @@ define(function (require, exports, module) {
             }
         });
     };
+    
+    /**
+     * This stores file system events emitted by watchers that were not yet processed
+     */
+    var _cachedFileSystemEvents = [];
+    
+    /**
+     * Debounced function to process emitted file system events
+     * for cases when there's a lot of fs events emitted in a very short period of time
+     */
+    _processCachedFileSystemEvents = _.debounce(function () {
+        // we need to reduce _cachedFileSystemEvents not to contain duplicates!
+        _cachedFileSystemEvents = _cachedFileSystemEvents.reduce(function (result, obj) {
+            var fullPath = obj.entry ? obj.entry.fullPath : null;
+            // merge added & removed
+            if (result[fullPath] && obj.isDirectory) {
+                obj.added = obj.added.concat(result[fullPath].added);
+                obj.removed = obj.removed.concat(result[fullPath].removed);
+            }
+            // use the latest event as base
+            result[fullPath] = obj;
+            return result;
+        }, {});
+        _.forEach(_cachedFileSystemEvents, function (obj) {
+            _fileSystemChangeHandler(obj.event, obj.entry, obj.added, obj.removed);
+        });
+        _cachedFileSystemEvents = [];
```

Comment:
`debounce` just wraps a function and keeps postponing its execution until the wrapper stops to be called for the timeout time. But once the wrapped function is executed, all happens totally sync so the code is safe. I've used this pattern a few times before.